plugins/core/hu.bme.mit.gamma.expression.model/src/hu/bme/mit/gamma/expression/util/ExpressionTypeDeterminator2.java

```diff
@@ -0,0 +1,403 @@
+import hu.bme.mit.gamma.expression.model.UnaryPlusExpression;
+import hu.bme.mit.gamma.expression.model.VoidTypeDefinition;
+import hu.bme.mit.gamma.util.GammaEcoreUtil;
+
+public class ExpressionTypeDeterminator2 {
+	public static final ExpressionTypeDeterminator2 INSTANCE = new ExpressionTypeDeterminator2();
+	protected ExpressionTypeDeterminator2() {}
+	
+	protected final GammaEcoreUtil ecoreUtil = GammaEcoreUtil.INSTANCE;
+	protected final ExpressionUtil expressionUtil = ExpressionUtil.INSTANCE;
+	protected final ExpressionModelFactory factory = ExpressionModelFactory.eINSTANCE;
+	
+	public Type getType(Expression expression) {
+		if (expression instanceof Declaration) {
+			return ExpressionModelDerivedFeatures.getTypeDefinition((Declaration)expression);
+		}
+		if (expression instanceof TypeReference) {
+			Type type = getType(expression);
+			for (TypeReference reference : ecoreUtil.getAllContentsOfType(type, TypeReference.class)) {
+				TypeDefinition typeDefinition = ExpressionModelDerivedFeatures.getTypeDefinition(reference);
+				ecoreUtil.replace(typeDefinition, reference);
+			}
+			return ExpressionModelDerivedFeatures.getTypeDefinition(type);
+		}
+		if (expression instanceof BooleanLiteralExpression) {
+			return factory.createBooleanTypeDefinition();
+		}
+		if (expression instanceof IntegerLiteralExpression) {
+			return factory.createIntegerTypeDefinition();
+		}
+		if (expression instanceof RationalLiteralExpression) {
+			return factory.createRationalTypeDefinition();
+		}
+		if (expression instanceof DecimalLiteralExpression) {
+			return factory.createDecimalTypeDefinition();
+		}
+		if (expression instanceof EnumerationLiteralExpression) {
+			return factory.createEnumerationTypeDefinition();
+		}
+		if (expression instanceof IntegerRangeLiteralExpression) {
+			return factory.createIntegerRangeTypeDefinition();
+		}
+		if (expression instanceof RecordLiteralExpression) {
+			return factory.createRecordTypeDefinition();
+		}
+		if (expression instanceof ArrayLiteralExpression) {
+			return factory.createArrayTypeDefinition();
+		}
+		if (expression instanceof DirectReferenceExpression) {
+			Type declarationType = ((DirectReferenceExpression) expression).getDeclaration().getType(); 
+			return ExpressionModelDerivedFeatures.getTypeDefinition(declarationType);
+		}
+		if (expression instanceof ElseExpression) {
+			return factory.createBooleanTypeDefinition();
+		}
+		if (expression instanceof BooleanExpression) {
+			Type declarationType = expressionUtil.getDeclaration(expression).getType();
+			return ExpressionModelDerivedFeatures.getTypeDefinition(declarationType);
+		}
+		if (expression instanceof PredicateExpression) {
+			return factory.createBooleanTypeDefinition();
+		}
+		if (expression instanceof QuantifierExpression) {
+			return factory.createBooleanTypeDefinition();
+		}
+		if (expression instanceof IfThenElseExpression) {
+			return factory.createBooleanTypeDefinition();
```

Comment:
Not true. Check the element of `.getThen()` and `.getElse()` (they must be the same).